src/main/java/com/charlesgadeken/entwined/EngineController.java

```diff
@@ -0,0 +1,146 @@
+package com.charlesgadeken.entwined;
+
+import com.charlesgadeken.entwined.effects.TSEffectController;
+import com.charlesgadeken.entwined.effects.original.ScrambleEffect;
+import com.charlesgadeken.entwined.effects.original.SpeedEffect;
+import com.charlesgadeken.entwined.effects.original.SpinEffect;
+import heronarts.lx.LX;
+import heronarts.lx.effect.BlurEffect;
+import heronarts.lx.mixer.LXAbstractChannel;
+import heronarts.lx.mixer.LXChannel;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class EngineController {
+    LX lx;
+
+    public int baseChannelIndex;
+    public int numChannels;
+
+    public int startEffectIndex;
+    public int endEffectIndex;
+
+    public boolean isAutoplaying;
+    EntwinedAutomationRecorder automation;
+    boolean[] previousChannelIsOn;
+
+    public List<TSEffectController> effectControllers = new ArrayList<>();
+    public int activeEffectControllerIndex = -1;
+
+    public SpeedEffect speedEffect;
+    public SpinEffect spinEffect;
+    public BlurEffect blurEffect;
+    public ScrambleEffect scrambleEffect;
+
+    public EngineController(LX lx) {
+        this.lx = lx;
+    }
+
+    public List<LXChannel> getChannels() {
+        // NOTE(meawoppl) dangercast here. @Slee?
```

Comment:
You could use LXAbstractChannel instead? Note that there are now LXGroup and LXChannel - an LXGroup can contain numerous other channels as a submix. So some channels from the engine may be LXGroup, not LXChannel. Both extend from LXAbstractChannel.

Reply:
This caller needs access to the current/next pattern indices?
https://github.com/squaredproject/Entwined/pull/16/files#diff-81e0a0a2175b3b6543ce59639959b4f4982324c987f5037460944b9b8eaaf823R27

The `LXAbstractChannel` does not seem to have those methods..

Reply:
Ah - right. Yes, this is because an LXGroup doesn't have patterns on it. It's just a container for a set of other channels. So these methods of selecting a pattern are only relevant to LXChannel.

In this case if I think rather than danger-casting what you want to do is filter the list of LXAbstractChannel from lx.engine.mixer and only return those that are of the concrete LXChannel type.

Reply:
This makes sense. Ty